File: src/shell/android/native_handler_thread.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "shell/android/native_handler_thread.h"

#include "base/message_loop/message_loop.h"
#include "jni/NativeHandlerThread_jni.h"

namespace shell {

jlong CreateMessageLoop(JNIEnv* env, const JavaParamRef<jobject>& jcaller) {
  scoped_ptr<base::MessageLoop> loop(new base::MessageLoopForUI);
  base::MessageLoopForUI::current()->Start();
  return reinterpret_cast<intptr_t>(loop.release());
}

void DeleteMessageLoop(JNIEnv* env,
                       const JavaParamRef<jobject>& jcaller,
                       jlong message_loop) {
  scoped_ptr<base::MessageLoop> loop(
      reinterpret_cast<base::MessageLoop*>(message_loop));
  loop->QuitNow();
}

bool RegisterNativeHandlerThreadJni(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

}  // namespace shell

File: src/shell/android/apk/src/org/chromium/mojo/shell/NativeHandlerThread.java
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.mojo.shell;

import android.os.HandlerThread;

import org.chromium.base.annotations.JNINamespace;

/**
 * Handler thread associated with a native message loop.
 */
@JNINamespace("shell")
public class NativeHandlerThread extends HandlerThread {
    // The native message loop.
    private long mNativeMessageLoop = 0;

    /**
     * Constructor.
     */
    public NativeHandlerThread(String name) {
        super(name);
    }

    /**
     * @see HandlerThread#run()
     */
    @Override
    public void run() {
        try {
            super.run();
        } finally {
            if (mNativeMessageLoop != 0) {
                nativeDeleteMessageLoop(mNativeMessageLoop);
            }
        }
    }

    /**
     * @see HandlerThread#onLooperPrepared()
     */
    @Override
    protected void onLooperPrepared() {
        if (mNativeMessageLoop == 0) {
            mNativeMessageLoop = nativeCreateMessageLoop();
        }
        super.onLooperPrepared();
    }

    native long nativeCreateMessageLoop();

    native void nativeDeleteMessageLoop(long messageLoop);
}
